extension-manifest-v3/src/background/adblocker.js

```diff
@@ -501,17 +491,35 @@ if (__PLATFORM__ === 'firefox') {
       }
 
       const request = Request.fromRequestDetails(details);
+      const cspPolicies = [];
+      const htmlFilters = [];
 
       if (!isTrusted(request, details.type)) {
         const engine = engines.get(engines.MAIN_ENGINE);
-        const policies = engine.getCSPDirectives(request);
 
-        if (policies !== undefined) {
-          return updateResponseHeadersWithCSP(details, policies);
+        htmlFilters.push(...engine.getHtmlFilters(request));
+
+        if (details.type === 'main_frame') {
+          const policies = engine.getCSPDirectives(request);
+          if (policies !== undefined) {
+            cspPolicies.push(...policies);
+          }
         }
       }
+
+      if (htmlFilters.length !== 0) {
+        filterRequestHTML(
+          chrome.webRequest.filterResponseData,
+          request,
+          htmlFilters,
+        );
+      }
+
+      if (cspPolicies.length !== 0) {
+        return updateResponseHeadersWithCSP(details, cspPolicies);
+      }
     },
-    { urls: ['<all_urls>'], types: ['main_frame'] },
+    { urls: ['http://*/*', 'https://*/*'] },
```

Comment:
Shouldn't we add the same constraint on other even listeners? I found a case where we have a callback for our own background page in Firefox (Navigation API) when Dev Tools are opened. 
Sometimes I have an error in Chrome, that shows that page from `chrome://` can't be accessed.